File: typescript/src/validation.ts
import { integer } from "./amounts.js";
import { definitions, requestContracts } from "./generated/requests.js";
import { assertSafeNumbers, isObject } from "./json.js";

interface Schema {
  const?: unknown;
  additionalProperties?: boolean | Schema;
  $ref?: string;
  type?: string | string[];
  enum?: unknown[];
  oneOf?: Schema[];
  anyOf?: Schema[];
  allOf?: Schema[];
  properties?: Record<string, Schema>;
  required?: string[] | boolean;
  items?: Schema;
  minimum?: number;
  maximum?: number;
  minItems?: number;
  maxItems?: number;
  minLength?: number;
  maxLength?: number;
  pattern?: string;
  format?: string;
}
const schemas = definitions as unknown as Record<string, Schema>;
function check(schema: Schema, value: unknown, path: string): void {
  if (schema.$ref) {
    const target = schemas[schema.$ref.split("/").at(-1) ?? ""];
    if (!target) throw new TypeError(`Unknown schema ${schema.$ref}`);
    check(target, value, path);
    return;
  }
  if (schema.oneOf || schema.anyOf) {
    let matches = 0;
    for (const choice of schema.oneOf ?? schema.anyOf ?? [])
      try {
        check(choice, value, path);
        matches++;
      } catch {}
    if (schema.oneOf ? matches !== 1 : matches < 1)
      throw new TypeError(`${path} does not match exactly one allowed shape`);
    return;
  }
  if ("const" in schema && value !== schema.const)
    throw new TypeError(`${path} differs from its constant`);
  for (const choice of schema.allOf ?? []) check(choice, value, path);
  const types = Array.isArray(schema.type) ? schema.type : [schema.type];
  if (value === null && types.includes("null")) return;
  if (schema.enum && !schema.enum.includes(value))
    throw new TypeError(`${path} has an unsupported value`);
  const type = types.find((t) => t !== "null");
  if (schema.type === "null" && value !== null) throw new TypeError(`${path} must be null`);
  if (type === "object") {
    if (!isObject(value)) throw new TypeError(`${path} must be an object`);
    for (const key of Object.keys(value))
      if (!schema.properties?.[key]) {
        if (schema.additionalProperties === false)
          throw new TypeError(`${path}.${key} is not permitted`);
        if (typeof schema.additionalProperties === "object")
          check(schema.additionalProperties, value[key], `${path}.${key}`);
      }
    for (const key of Array.isArray(schema.required) ? schema.required : [])
      if (value[key] === undefined) throw new TypeError(`${path}.${key} is required`);
    for (const [key, s] of Object.entries(schema.properties ?? {}))
      if (value[key] !== undefined) check(s, value[key], `${path}.${key}`);
  } else if (type === "array") {
    if (!Array.isArray(value)) throw new TypeError(`${path} must be an array`);
    if (value.length < (schema.minItems ?? 0) || value.length > (schema.maxItems ?? Infinity))
      throw new RangeError(`${path} has invalid length`);
    if (schema.items) for (const item of value) check(schema.items, item, path);
  } else if (type === "string") {
    if (typeof value !== "string") throw new TypeError(`${path} must be a string`);
    if (
      schema.format === "int64" &&
      (!/^-?\d+$/.test(value) || BigInt(value) < -(1n << 63n) || BigInt(value) > (1n << 63n) - 1n)
    )
      throw new TypeError(`${path} is not a signed 64-bit decimal string`);
    if (
      value.length < (schema.minLength ?? 0) ||
      value.length > (schema.maxLength ?? Infinity) ||
      (schema.pattern && !new RegExp(schema.pattern).test(value))
    )
      throw new TypeError(`${path} has invalid format`);
  } else if (type === "integer" || type === "number") {
    if (typeof value !== "number" && typeof value !== "bigint")
      throw new TypeError(`${path} must be a number or bigint`);
    if (type === "integer" && typeof value === "number" && !Number.isSafeInteger(value))
      throw new TypeError(`${path} must be an exact integer`);
    if (
      (schema.minimum !== undefined && value < schema.minimum) ||
      (schema.maximum !== undefined && value > schema.maximum)
    )
      throw new RangeError(`${path} is outside its range`);
  } else if (type === "boolean" && typeof value !== "boolean")
    throw new TypeError(`${path} must be boolean`);
}
function micro(value: unknown, name: string, maximum = (1n << 63n) - 1n): bigint {
  if (typeof value !== "string" && typeof value !== "bigint" && typeof value !== "number")
    throw new TypeError(`${name} is required`);
  return integer(value, name, maximum, 1n);
}
export function validateOrder(value: unknown, signed = false): void {
  if (!isObject(value)) throw new TypeError("Order must be an object");
  if (value.side !== "BUY" && value.side !== "SELL")
    throw new TypeError("side must be BUY or SELL");
  if (!["GTC", "GTD", "FAK", "FOK"].includes(String(value.time_in_force)))
    throw new TypeError("Invalid time_in_force");
  if (value.type === "LIMIT") {
    micro(value.price_micro, "price_micro", 999999n);
    micro(value.shares_micro, "shares_micro");
    if (value.collateral_amount_micro != null)
      throw new TypeError("LIMIT orders require shares_micro, not collateral_amount_micro");
  } else if (value.type === "MARKET" && !signed) {
    if (value.price_micro != null || !["FAK", "FOK"].includes(String(value.time_in_force)))
      throw new TypeError("MARKET orders require FAK/FOK and no price");
    micro(
      value.side === "BUY" ? value.collateral_amount_micro : value.shares_micro,
      "market amount",
    );
    if (value.side === "BUY" ? value.shares_micro != null : value.collateral_amount_micro != null)
      throw new TypeError("MARKET BUY needs collateral, SELL needs shares");
  } else throw new TypeError("Unsupported order type");
  if (value.time_in_force === "GTD") {
    if (
      typeof value.expiration_unix_seconds !== "number" ||
      !Number.isSafeInteger(value.expiration_unix_seconds) ||
      value.expiration_unix_seconds <= Date.now() / 1000
    )
      throw new TypeError("GTD needs a future expiration");
  } else if (value.expiration_unix_seconds != null) throw new TypeError("Expiration requires GTD");
  if (value.post_only && !["GTC", "GTD"].includes(String(value.time_in_force)))
    throw new TypeError("post_only requires GTC/GTD");
  if (signed) {
    if (value.side_u8 !== (value.side === "BUY" ? 0 : 1))
      throw new TypeError("Signed side does not match order side");
    const shares = micro(value.shares_micro, "shares_micro"),
      price = micro(value.price_micro, "price_micro", 999999n),
      collateral = (shares * price) / 1000000n;
    if (
      BigInt(String(value.maker_amount)) !== (value.side === "BUY" ? collateral : shares) ||
      BigInt(String(value.taker_amount)) !== (value.side === "BUY" ? shares : collateral)
    )
      throw new TypeError("Signed amounts do not match order economics");
    if (BigInt(String(value.token_id)) !== BigInt(String(value.chain_token_id)))
      throw new TypeError("Signed token does not match order token");
    integer(String(value.salt), "salt", (1n << 256n) - 1n, 1n);
    if (
      !/^0x[0-9a-fA-F]{64}$/.test(String(value.order_hash)) ||
      !/^0x[0-9a-fA-F]{130}$/.test(String(value.signature))
    )
      throw new TypeError("Malformed order hash or signature");
  }
}
export function validateRequest(name: string, body: unknown, query: unknown): void {
  assertSafeNumbers(body);
  assertSafeNumbers(query);
  const contract = (
    requestContracts as unknown as Record<string, { body?: Schema; query: Record<string, Schema> }>
  )[name];
  if (contract?.body) check(contract.body, body, "body");
  if (contract)
    for (const [key, schema] of Object.entries(contract.query)) {
      const value = isObject(query) ? query[key] : undefined;
      if (value === undefined) {
        if (schema.required === true) throw new TypeError(`query.${key} is required`);
      } else check(schema, value, `query.${key}`);
    }
  if (name === "placeOrder" || name === "placeSignedOrder")
    validateOrder(body, name === "placeSignedOrder");
  if (name === "placeSignedOrders") {
    if (
      !isObject(body) ||
      !Array.isArray(body.orders) ||
      body.orders.length < 1 ||
      body.orders.length > 32
    )
      throw new RangeError("Signed order batches need 1–32 entries");
    for (const order of body.orders) validateOrder(order, true);
  }
  if ((name === "splitPosition" || name === "mergePosition") && isObject(body))
    micro(body.collateral_amount_micro ?? body.shares_micro, "position amount");
  if ((name === "submitBatch" || name === "supersedeBatch") && isObject(body)) {
    if (!Array.isArray(body.ops) || body.ops.length < 1 || body.ops.length > 20)
      throw new RangeError("Account batches need 1–20 operations");
    for (const op of body.ops) {
      if (!isObject(op) || !["SPLIT", "MERGE", "WITHDRAW"].includes(String(op.kind)))
        throw new TypeError("Presigned batches support SPLIT, MERGE, and WITHDRAW only");
      micro(op.kind === "WITHDRAW" ? op.amount_micro : op.shares_micro, "batch amount");
    }
    if (!/^0x[0-9a-fA-F]{130}$/.test(String(body.signature)))
      throw new TypeError("Expected a 65-byte batch signature");
    if (name === "submitBatch") integer(String(body.seq), "seq", (1n << 63n) - 1n);
    integer(String(body.deadline_unix_seconds), "deadline", (1n << 63n) - 1n);
  }
}

export function validateResponse(name: string, status: number, value: unknown): void {
  const schema = (
    requestContracts as unknown as Record<string, { responses: Record<string, Schema> }>
  )[name]?.responses[String(status)];
  if (schema) check(schema, value, "response");
}

File: rust/src/validation.rs
//! Machine-matchable validation failures shared by request boundaries.

mod reasons;

pub use reasons::{AmountReason, IdentifierReason, OrderInputReason, QueryReason};

/// The precise request or configuration field rejected before I/O.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum Field {
	/// Bridge deposit source-chain identifier.
	FromChainId,

	/// Bridge withdrawal destination-chain identifier.
	ToChainId,

	/// Provider-specific source token address.
	FromTokenAddress,

	/// Provider-specific destination token address.
	ToTokenAddress,

	/// Provider-specific bridge recipient address.
	RecipientAddress,

	/// Positive integer amount in a token’s native base units.
	BaseUnitAmount,

	/// An identifier whose role is chosen by a generic boundary.
	Identifier,

	/// Account-batch group UUID.
	BatchGroupId,

	/// HTTP request path.
	Path,

	/// Personal access token credential.
	ApiToken,

	/// Order collateral budget.
	Collateral,

	/// A single exchange selection.
	Exchange,

	/// Market lifecycle-state filter.
	State,

	/// Parent event slug filter.
	EventSlug,

	/// Compound query configuration.
	Query,

	/// HTTP service origin and optional reverse-proxy path.
	BaseUrl,

	/// Bearer credential used for private endpoints.
	Token,

	/// Overall request timeout.
	Timeout,

	/// Connection establishment timeout.
	ConnectTimeout,

	/// Response read timeout.
	ReadTimeout,

	/// Maximum duration of a polling operation.
	PollTimeout,

	/// Delay between status reads.
	PollInterval,

	/// First retry delay.
	InitialBackoff,

	/// Upper bound on retry delay.
	MaxBackoff,

	/// Fraction of randomized retry delay.
	Jitter,

	/// Maximum number of retries.
	MaxRetries,

	/// Outcome token identifier.
	TokenId,

	/// On-chain market condition identifier.
	ConditionId,

	/// Internal order UUID.
	OrderId,

	/// Signed order digest.
	OrderHash,

	/// Internal wallet UUID.
	WalletId,

	/// Signed account-batch digest.
	BatchHash,

	/// Account-batch group UUID.
	GroupId,

	/// Catalogue market UUID.
	MarketId,

	/// Catalogue event UUID.
	EventId,

	/// EVM account or contract address.
	Address,

	/// Encoded cryptographic signature.
	Signature,

	/// Order uniqueness nonce.
	Salt,

	/// Order metadata digest.
	Metadata,

	/// Order builder digest.
	Builder,

	/// Signing domain chain identifier.
	ChainId,

	/// Account sequence bound by a batch signature.
	Sequence,

	/// Batch signature deadline.
	Deadline,

	/// Price expressed in whole collateral units.
	Price,

	/// Quantity expressed in whole shares.
	Shares,

	/// Budget expressed in whole collateral units.
	CollateralAmount,

	/// Price in integer micro units.
	PriceMicro,

	/// Quantity in integer micro shares.
	SharesMicro,

	/// Budget in integer micro collateral.
	CollateralAmountMicro,

	/// Whole-unit amount being converted.
	Amount,

	/// Integer micro-unit amount.
	AmountMicro,

	/// Order buy or sell direction.
	Side,

	/// Limit or market order kind.
	OrderType,

	/// Order persistence policy.
	TimeInForce,

	/// Maker-only execution flag.
	PostOnly,

	/// Good-til-date expiration timestamp.
	Expiration,

	/// Collection of independently signed orders.
	Orders,

	/// Account-batch operations.
	Operations,

	/// Requested exchange filters.
	Exchanges,

	/// Requested market condition filters.
	ConditionIds,

	/// Requested outcome filters.
	TokenIds,

	/// Maximum entries in a result page.
	Limit,

	/// Opaque pagination continuation token.
	Cursor,

	/// One-based page number.
	Page,

	/// Catalogue or incentive category filter.
	Category,

	/// Literal incentive-market search text.
	Search,

	/// Field selected for incentive sorting.
	SortBy,

	/// Ascending or descending sort direction.
	SortOrder,

	/// Inclusive range start.
	From,

	/// Inclusive range end.
	To,

	/// Timestamp of a historical price lookup.
	At,

	/// Calendar trading date.
	Date,

	/// Trading venue market identifier code.
	Mic,

	/// Canonical security or provider symbol.
	Symbol,

	/// Price data provider identifier.
	Provider,

	/// Token-price history window.
	Range,

	/// Number of sampled price points.
	Points,

	/// Realized-PnL bucket resolution.
	Granularity,

	/// Realized-PnL reporting window.
	Window,

	/// Catalogue exchange source.
	Source,

	/// Catalogue resource slug.
	Slug,

	/// Root category hierarchy filter.
	Root,

	/// Games or propositions event filter.
	EventTypeBucket,

	/// Named catalogue event filter.
	Filter,

	/// Whether ended events are excluded.
	ExcludeEnded,

	/// Event resolution filter.
	Resolution,

	/// Event ordering policy.
	Sort,

	/// Whether catalogue market enrichment is requested.
	IncludeMarkets,

	/// A single resource identifier in an HTTP path.
	PathSegment,
}

/// A rejected domain identifier and the violated invariant.
#[derive(Debug, thiserror::Error)]
#[error("invalid {field:?}: {reason}")]
pub struct IdentifierError {
	/// Typed identifier field that failed validation.
	pub field: Field,

	/// Specific identifier invariant that was violated.
	#[source]
	pub reason: IdentifierReason,
}

/// An amount that cannot be represented or accepted without rounding.
#[derive(Debug, thiserror::Error)]
#[error("invalid {field:?}: {reason}")]
pub struct AmountError {
	/// Amount field whose conversion or range check failed.
	pub field: Field,

	/// Exact conversion or amount-range failure.
	#[source]
	pub reason: AmountReason,
}

/// A query constraint rejected before a request is built.
#[derive(Debug, thiserror::Error)]
#[error("invalid {field:?}: {reason}")]
pub struct QueryError {
	/// Query parameter that requires correction.
	pub field: Field,

	/// Machine-matchable constraint, including conflicting fields when relevant.
	#[source]
	pub reason: QueryReason,
}

/// Invalid relationships among unsigned order inputs.
#[derive(Debug, thiserror::Error)]
#[error("invalid {field:?}: {reason}")]
pub struct OrderInputError {
	/// Order field that requires correction.
	pub field: Field,

	/// Violated order-shape or lifetime constraint.
	#[source]
	pub reason: OrderInputReason,
}

/// Invalid HTTP client configuration; parser errors retain their original source.
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum ConfigurationError {
	/// The configured service URL could not be parsed.
	#[error("invalid base URL: {0}")]
	InvalidBaseUrl(#[source] url::ParseError),

	/// The bearer token cannot be represented as an HTTP header value.
	#[error("invalid token header: {0}")]
	InvalidTokenHeader(#[source] reqwest::header::InvalidHeaderValue),

	/// A private client was configured with an empty bearer credential.
	#[error("token must be nonempty")]
	EmptyToken,

	/// Only HTTP and HTTPS service origins are supported.
	#[error("base URL must use HTTP or HTTPS")]
	UnsupportedScheme,

	/// User information in a service URL would introduce a second authentication mechanism.
	#[error("base URL must not contain credentials")]
	UrlCredentials,

	/// Query parameters belong on individual requests.
	#[error("base URL must not contain a query")]
	UrlQuery,

	/// URL fragments are not sent to an HTTP server.
	#[error("base URL must not contain a fragment")]
	UrlFragment,

	/// The service URL has no network host.
	#[error("base URL must contain a host")]
	MissingHost,

	/// A configured timeout cannot provide a finite, positive deadline.
	#[error("{field:?} must be finite and positive")]
	InvalidTimeout {
		/// Timeout setting that must be corrected.
		field: Field,
	},
}

/// An invalid retry policy; no retry is scheduled when validation fails.
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum RetryPolicyError {
	/// Retry delays must not create a tight loop.
	#[error("initial retry backoff must be positive")]
	ZeroInitialBackoff,

	/// A backoff ceiling cannot be smaller than the initial delay.
	#[error("maximum backoff must be at least the initial backoff")]
	MaxBelowInitial,

	/// Jitter must be finite and between zero and one.
	#[error("retry jitter must be finite and in [0, 1]")]
	InvalidJitter,

	/// The configured attempt count exceeds the supported policy limit.
	#[error("retry count exceeds the supported maximum")]
	ExcessiveAttempts,
}

/// All shared local input-validation families.
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum ValidationError {
	/// A resource identifier violates its format or size constraints.
	#[error(transparent)]
	Identifier(#[from] IdentifierError),

	/// An amount is invalid, inexact, or outside the supported range.
	#[error(transparent)]
	Amount(#[from] AmountError),

	/// HTTP transport configuration cannot be used safely.
	#[error(transparent)]
	Configuration(#[from] ConfigurationError),

	/// Query parameters violate a documented endpoint constraint.
	#[error(transparent)]
	Query(#[from] QueryError),

	/// Unsigned order fields form an unsupported combination.
	#[error(transparent)]
	Order(#[from] OrderInputError),

	/// A retry policy cannot provide a bounded backoff.
	#[error(transparent)]
	RetryPolicy(#[from] RetryPolicyError),
}

impl ValidationError {
	/// Retarget a field-bearing failure without changing its category or original parser source.
	/// Configuration and retry errors without a field retain their intrinsic identity.
	pub fn with_field(mut self, field: Field) -> Self {
		match &mut self {
			Self::Identifier(error) => error.field = field,
			Self::Amount(error) => error.field = field,
			Self::Query(error) => error.field = field,
			Self::Order(error) => error.field = field,
			Self::Configuration(ConfigurationError::InvalidTimeout { field: original }) => {
				*original = field
			},
			Self::Configuration(_) | Self::RetryPolicy(_) => {},
		}

		self
	}

	/// Reject an identifier while preserving its field and exact invariant.
	pub const fn identifier(field: Field, reason: IdentifierReason) -> Self {
		Self::Identifier(IdentifierError { field, reason })
	}

	/// Reject an amount without substituting a rounded or saturated value.
	pub const fn amount(field: Field, reason: AmountReason) -> Self {
		Self::Amount(AmountError { field, reason })
	}

	/// Reject a query parameter before issuing network I/O.
	pub const fn query(field: Field, reason: QueryReason) -> Self {
		Self::Query(QueryError { field, reason })
	}

	/// Reject an invalid unsigned order field combination.
	pub const fn order(field: Field, reason: OrderInputReason) -> Self {
		Self::Order(OrderInputError { field, reason })
	}
}
